File: zenwave-jpa-example/src/main/java/io/zenwave360/example/adapters/web/CustomerApiController.java
package io.zenwave360.example.adapters.web;

import io.zenwave360.example.adapters.web.mappers.CustomerDTOsMapper;
import io.zenwave360.example.adapters.web.model.CustomerDTO;
import io.zenwave360.example.adapters.web.model.CustomerPaginatedDTO;
import io.zenwave360.example.adapters.web.model.CustomerSearchCriteriaDTO;
import io.zenwave360.example.core.inbound.CustomerService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;
import org.springframework.web.context.request.NativeWebRequest;

import java.util.List;
import java.util.Optional;

/** REST controller for CustomerApi. */
@RestController
@RequestMapping("/api")
public class CustomerApiController implements CustomerApi {

    private final Logger log = LoggerFactory.getLogger(getClass());

    @Autowired
    private NativeWebRequest request;

    private CustomerService customerService;

    @Autowired
    public CustomerApiController setCustomerService(CustomerService customerService) {
        this.customerService = customerService;
        return this;
    }

    private CustomerDTOsMapper mapper = CustomerDTOsMapper.INSTANCE;

    public CustomerApiController(CustomerService customerService) {

        this.customerService = customerService;
    }

    public Optional<NativeWebRequest> getRequest() {
        return Optional.ofNullable(request);
    }

    @Override
    public ResponseEntity<CustomerDTO> createCustomer(CustomerDTO reqBody) {
        log.debug("REST request to createCustomer: {}", reqBody);
        var input = mapper.asCustomer(reqBody);
        var customer = customerService.createCustomer(input);
        CustomerDTO responseDTO = mapper.asCustomerDTO(customer);
        return ResponseEntity.status(201).body(responseDTO);
    }

    @Override
    public ResponseEntity<CustomerDTO> getCustomer(Long id) {
        log.debug("REST request to getCustomer: {}", id);
        var customer = customerService.getCustomer(id);
        if (customer.isPresent()) {
            CustomerDTO responseDTO = mapper.asCustomerDTO(customer.get());
            return ResponseEntity.status(200).body(responseDTO);
        }
        else {
            return ResponseEntity.notFound().build();
        }
    }

    @Override
    public ResponseEntity<CustomerDTO> updateCustomer(Long id, CustomerDTO reqBody) {
        log.debug("REST request to updateCustomer: {}, {}", id, reqBody);
        var input = mapper.asCustomer(reqBody);
        var customer = customerService.updateCustomer(id, input);
        if (customer.isPresent()) {
            CustomerDTO responseDTO = mapper.asCustomerDTO(customer.get());
            return ResponseEntity.status(200).body(responseDTO);
        }
        else {
            return ResponseEntity.notFound().build();
        }
    }

    @Override
    public ResponseEntity<Void> deleteCustomer(Long id) {
        log.debug("REST request to deleteCustomer: {}", id);
        customerService.deleteCustomer(id);
        return ResponseEntity.status(204).build();
    }

    @Override
    public ResponseEntity<CustomerPaginatedDTO> searchCustomers(
            CustomerSearchCriteriaDTO reqBody,
            Optional<Integer> page, Optional<Integer> limit, Optional<List<String>> sort) {
        log.debug("REST request to searchCustomers: {}, {}, {}, {}", page, limit, sort, reqBody);
        var input = mapper.asCustomerSearchCriteria(reqBody);
        var customerPage = customerService.searchCustomers(input, pageOf(page, limit, sort));
        var responseDTO = mapper.asCustomerPaginatedDTO(customerPage);
        return ResponseEntity.status(200).body(responseDTO);
    }

    protected Pageable pageOf(Optional<Integer> page, Optional<Integer> limit, Optional<List<String>> sort) {
        Sort sortOrder = sort.map(s -> Sort.by(s.stream().map(sortParam -> {
            String[] parts = sortParam.split(":");
            String property = parts[0];
            Sort.Direction direction = parts.length > 1 ? Sort.Direction.fromString(parts[1]) : Sort.Direction.ASC;
            return new Sort.Order(direction, property);
        }).toList())).orElse(Sort.unsorted());
        return PageRequest.of(page.orElse(0), limit.orElse(10), sortOrder);
    }

}

File: examples/kustomer-address-jpa/src/main/kotlin/io/zenwave360/examples/kotlin/web/CustomerApiController.kt
package io.zenwave360.examples.kotlin.web

import io.zenwave360.examples.kotlin.domain.*
import io.zenwave360.examples.kotlin.service.*
import io.zenwave360.examples.kotlin.service.dtos.*
import io.zenwave360.examples.kotlin.web.*
import io.zenwave360.examples.kotlin.web.dtos.*
import io.zenwave360.examples.kotlin.web.mappers.*

import java.net.URI
import java.net.URISyntaxException
import java.math.*
import java.time.*
import java.util.*
import jakarta.validation.Valid
import jakarta.validation.constraints.NotNull
import org.mapstruct.factory.Mappers
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.core.io.ByteArrayResource
import org.springframework.core.io.Resource
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageRequest
import org.springframework.data.domain.Pageable
import org.springframework.data.domain.Sort
import org.springframework.web.context.request.NativeWebRequest


/**
 * REST controller for CustomerApi.
 */
@RestController
@RequestMapping("/api")
open class CustomerApiController(
    private val customerService: CustomerService
) : CustomerApi {

    private val log: Logger = LoggerFactory.getLogger(javaClass)

    @Autowired
    private lateinit var request: NativeWebRequest

    private val mapper = CustomerDTOsMapper.INSTANCE

    fun getRequest(): Optional<NativeWebRequest> {
        return Optional.ofNullable(request)
    }



    override fun createCustomer(reqBody: CustomerDTO): ResponseEntity<CustomerDTO> {
        log.debug("REST request to createCustomer: {}", reqBody)
        val input = mapper.asCustomer(reqBody)
        val customer =  customerService.createCustomer(input)
        val responseDTO = mapper.asCustomerDTO(customer)
        return ResponseEntity.status(201).body(responseDTO)
    }

    override fun getCustomer(customerId: Long): ResponseEntity<CustomerDTO> {
        log.debug("REST request to getCustomer: {}", customerId)
        val customer =  customerService.getCustomer(customerId)
        return if (customer.isPresent) {
            val responseDTO = mapper.asCustomerDTO(customer.get())
            ResponseEntity.status(200).body(responseDTO)
        } else {
            ResponseEntity.notFound().build()
        }
    }

    override fun updateCustomer(customerId: Long, reqBody: CustomerDTO): ResponseEntity<CustomerDTO> {
        log.debug("REST request to updateCustomer: {}, {}", customerId, reqBody)
        val input = mapper.asCustomer(reqBody)
        val customer =  customerService.updateCustomer(customerId, input)
        return if (customer.isPresent) {
            val responseDTO = mapper.asCustomerDTO(customer.get())
            ResponseEntity.status(200).body(responseDTO)
        } else {
            ResponseEntity.notFound().build()
        }
    }

    override fun deleteCustomer(customerId: Long): ResponseEntity<Unit> {
        log.debug("REST request to deleteCustomer: {}", customerId)
        customerService.deleteCustomer(customerId)
        return ResponseEntity.status(204).build()
    }

    override fun searchCustomers(page: Int, limit: Int, sort: List<String>?, reqBody: CustomerSearchCriteriaDTO): ResponseEntity<CustomerPaginatedDTO> {
        log.debug("REST request to searchCustomers: {}, {}, {}, {}", page, limit, sort, reqBody)
        val input = mapper.asCustomerSearchCriteria(reqBody)
        val customerPage =  customerService.searchCustomers(input, pageOf(page, limit, sort))
        val responseDTO = mapper.asCustomerPaginatedDTO(customerPage)
        return ResponseEntity.status(200).body(responseDTO)
    }


    protected fun pageOf(page: Int?, limit: Int?, sort: List<String>?): Pageable {
        val sortOrder = sort?.let {
            Sort.by(it.map { sortParam ->
                val parts = sortParam.split(":")
                val property = parts[0]
                val direction = if (parts.size > 1) Sort.Direction.fromString(parts[1]) else Sort.Direction.ASC
                Sort.Order(direction, property)
            })
        } ?: Sort.unsorted()
        return PageRequest.of(page ?: 0, limit ?: 10, sortOrder)
    }
}
